lib/services/local/dynamic_access.go

```diff
@@ -372,19 +411,25 @@ func (s *DynamicAccessService) updateAccessRequestPluginData(ctx context.Context
 				return trace.Wrap(err)
 			}
 			// In order to prevent orphaned plugin data, we automatically
-			// configure new instances to expire shortly after the AccessRequest
+			// configure new instances to expire shortly after the resource
 			// to which they are associated.  This discrepency in expiry gives
 			// plugins the ability to use stored data when handling an expiry
 			// (OpDelete) event.
-			req, err := s.GetAccessRequest(ctx, params.Resource)
+			exp, err := s.getResourceExpiry(ctx, params.Resource, params.Kind)
 			if err != nil {
 				return trace.Wrap(err)
 			}
-			data, err = types.NewPluginData(params.Resource, types.KindAccessRequest)
+			data, err = types.NewPluginData(params.Resource, params.Kind)
 			if err != nil {
 				return trace.Wrap(err)
 			}
-			data.SetExpiry(req.GetAccessExpiry().Add(time.Hour))
+
+			// A resource can have empty expiration time by default.
+			// In this case, we should leave PluginData expiration time empty as well.
+			// TODO: PluginData must be destroyed with parent resource. This needs to be addressed in future.
+			if (exp != time.Time{}) {
```

Comment:
Would...

```suggestion
			if (exp.IsZero()) {
```
... work here?

Reply:
Absolutely. Thank you, fixed.

Reply:
Oops - forgto an `!`

Reply:
> @gzigzigzeo Can you help me understand the use case for separating out plugin data? Is this for the event-handler login locking?

Not exactly. User locking data is now in-memory. I do not see a lot of benefits of persisting it because `LockPeriod` is relatively short. Loosing it wouldn't hurt much. In the another hand, persisting it would complicate things a bit. 

The motivation behind this is the following:

1. Get rid of disk state in `event-handler`. For now, it is very compact and consists of a bunch of cursors (strings and ints). Disk turned out to be unreliable storage, especially in k8s. As far as I understood, clients are experiencing problems. Sometimes it gets corrupted and lost. Moving this state to the API would solve the reliability problems.
2. We are going to need to provide some kind of a simple persistent storage to our feature WASM plugin authors.
3. As far as I know, this would potentially be useful for other plugins we have (@marshall-lee might add his thoughts here).

I have separated `PluginData` because it was too tightly coupled to `AccessRequest` and I needed to expand it to work with users.

> 1. The user doesn't need access to plugin data, only the plugin user right? In which case the plugin user's role will need an allow rule for `KindPluginData` and verbs `read,list,update,delete`.

That's right. I think, that would be enough for now. Plugin user should have access to their own plugin data, admin user should be able to access a plugin data of any user. I think, ideally, we would want to make access configureable using labels.

> 2. That's a good point and probably should be solved in this Pr, not sure how yet

I think that for now we could introduce a simple workaround which sets `PluginData` TTL to Now()+1h when user gets deleted. 

But I also think that we need a universal way of managing resource TTLs and relations. The backend service seems the good candidate to put this responsibility on. However, I am not yet sure how exactly.

Reply:
No worries, it has been moved to another file, and ! is in place there :-)